### PR TITLE
Add TLS support

### PR DESCRIPTION
**Pull request by sneginhal**

### CODE REVIEW

**Comment by slfritchie** [comment]
Here is the recipe that I used to check that data "on the wire" appears to be encrypted.  At the top of the CorfuDB source repo:

    echo single-password > ./tmp.single-password-pw-file
    keytool -keystore ./tmp.singlestore -genkey -alias single1 -keyalg RSA -dname "cn=localhost, ou=IT, o=Continuent, c=US" -storepass `cat ./tmp.single-password-pw-file` -keypass `cat ./tmp.single-password-pw-file`
    ./bin/corfu_server -ms -e -u ./tmp.singlestore -f ./tmp.single-password-pw-file -r ./tmp.singlestore -w ./tmp.single-password-pw-file 9000 &
    sleep 3
    ./bin/corfu_query -e -u ./tmp.singlestore -f ./tmp.single-password-pw-file -r ./tmp.singlestore -w ./tmp.single-password-pw-file localhost:9000

The `corfu_query` command returns the Java object rendering of the JSON query response, no errors are reported by client or server, and a `tcpdump` capture of traffic on TCP port 9000 across the loopback interface contains no cleartext of the query or of the JSON response.

Now that I've figured out how to do this, my question earlier today to @sneginhal can be ignored, thanks.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 1]
```diff
@@ -183,6 +212,100 @@ public static void main(String[] args) {
         addManagementServer();
         router.baseServer.setOptionsMap(opts);
 
+        // Setup SSL if needed
+        Boolean tlsEnabled = (Boolean) opts.get("--enable-tls");
+        if (tlsEnabled) {
+            // Get the key store password
```
Comment (slfritchie) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Any opinions for or against that small refactoring?  Aside from that question, everything else looks good to me.  As an added bonus, when TLS is enabled, ciphertext does indeed appear to be on the wire, and also attempts to have a non-TLS client to talk to a TLS-enabled server will fail.

Comment (sneginhal) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Thanks for looking at this PR Scott.
+1 for the refactoring.
Here are the key stores and trust stores I used for manual test:
https://github.com/CorfuDB/CorfuDB/compare/76f9ed5...CorfuDB:tls-support

File: test/src/test/resources/password [hunk 2]
```diff
@@ -0,0 +1 @@
+cashcow
```
Comment (no2chem) [inline test/src/test/resources/password]: I recommend changing this to "test"

Comment (sneginhal) [inline test/src/test/resources/password]: Done. Changed the password to test123 on the stores.

Writing the doc.

Comment (sneginhal) [inline test/src/test/resources/password]: https://github.com/CorfuDB/CorfuDB/blob/tls-support/TLS.md

File: TLS.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+Run corfu_server
+```bash
+$ bin/corfu_server -m -e -u s1.jks -f s1pass.txt -r trust.jks -w tpass.txt --single 9000
+```
+
+Runtime:
+
+```java
+    CorfuRuntime rt = new CorfuRuntime("localhost:9000");
+    r.enableTls("r1.jks", "r1pass.txt", "trust.jks", "tpass.txt");
```
Comment (no2chem) [inline TLS.md]: do you mean rt here?

File: runtime/src/main/java/org/corfudb/util/TlsUtils.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        // Get the trust store
+        KeyStore ts = null;
+        if (trustStore != null) {
+            try (FileInputStream fis = new FileInputStream(trustStore)) {
+                ts = KeyStore.getInstance(KeyStore.getDefaultType());
+                ts.load(fis, tsp.toCharArray());
+            } catch (Exception e) {
+                tsPasswordFileException.accept(e);
+            }
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/util/TlsUtils.java]: not sure why this does not return null on exception.  The parameters did.

File: runtime/src/main/java/org/corfudb/util/TlsUtils.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Created by sfritchie on 1/12/17.
+ */
+public class TlsUtils {
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/util/TlsUtils.java]: Comments please.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 6]
```diff
@@ -183,6 +213,54 @@ public static void main(String[] args) {
...
+                List<String> ciphers = Pattern.compile(",")
+                    .splitAsStream(ciphs)
+                    .map(String::trim)
+                    .collect(Collectors.toList());
+                enabledTlsCipherSuites = ciphers.toArray(new String[ciphers.size()]);
+            }
+
+            // Get the TLS protocols to enable
+            String protos = (String) opts.get("--tls-protocols");
+            if (protos != null) {
```
Comment (sneginhal) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: If this option is not set, opts should have the default value of TLSv1.1,TLSv1.2. So protos should not be null.
If it is null at this point for some reason, enabledTlsProtocols will be null and the later call to setEnabledProtocols() will throw an IllegalArgumentException, which is what would happen even if the user sets this option to some junk.

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 7]
```diff
@@ -13,22 +13,37 @@
...
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.util.List;
 import java.util.Map;
-import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.ThreadFactory;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.net.ssl.*;
```
Comment (sneginhal) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: Done.

File: runtime/src/main/java/org/corfudb/util/TlsUtils.java (hunk 4 above)
Comment (sneginhal) [inline runtime/src/main/java/org/corfudb/util/TlsUtils.java]: Good catch, done.

**Comment by sneginhal** [comment]
Hi Michael,

Can you please merge this PR? All review comments have been addressed.

**Comment by no2chem** [comment]
merged, thanks!

**Pull request merged**